src/function/algorithm/wekaly_connected_component.cpp

```diff
@@ -0,0 +1,116 @@
+        nodeIDVector->state = DataChunkState::getSingleValueDataChunkState();
+        groupVector->state = DataChunkState::getSingleValueDataChunkState();
+        vectors.push_back(nodeIDVector.get());
+        vectors.push_back(groupVector.get());
+    }
+
+    void materialize(graph::Graph* graph, const std::vector<int64_t>& groupArray,
+        FactorizedTable& table) const {
+        for (auto offset = 0u; offset < graph->getNumNodes(); ++offset) {
+            nodeIDVector->setValue<nodeID_t>(0, {offset, graph->getNodeTableID()});
+            groupVector->setValue<int64_t>(0, groupArray[offset]);
+            table.append(vectors);
+        }
+    }
+
+private:
+    std::unique_ptr<ValueVector> nodeIDVector;
+    std::unique_ptr<ValueVector> groupVector;
+    std::vector<ValueVector*> vectors;
+};
+
+class WeaklyConnectedComponent final : public GDSAlgorithm {
+public:
+    WeaklyConnectedComponent() = default;
+    WeaklyConnectedComponent(const WeaklyConnectedComponent& other) : GDSAlgorithm{other} {}
+
+    std::vector<common::LogicalTypeID> getParamTypeIDs() const override {
+        return std::vector<LogicalTypeID>{LogicalTypeID::ANY};
+    }
+
+    std::vector<std::string> getResultColumnNames() const override {
+        return {"node_id", "group_id"};
+    }
+
+    std::vector<common::LogicalType> getResultColumnTypes() const override {
+        return {*LogicalType::INTERNAL_ID(), *LogicalType::INT64()};
+    }
+
+    void initLocalState(main::ClientContext* context) override {
+        localState = std::make_unique<WeaklyConnectedComponentLocalState>(context);
+    }
+
+    void exec() override {
+        auto wccLocalState = localState->ptrCast<WeaklyConnectedComponentLocalState>();
+        // Initialize state.
+        visitedArray.resize(graph->getNumNodes());
+        groupArray.resize(graph->getNumNodes());
+        for (auto offset = 0u; offset < graph->getNumNodes(); ++offset) {
+            visitedArray[offset] = false;
+            groupArray[offset] = -1;
+        }
+        // Compute weakly connected component.
+        auto groupID = 0;
+        for (auto offset = 0u; offset < graph->getNumNodes(); ++offset) {
+            if (visitedArray[offset]) {
+                continue;
+            }
+            findConnectedComponent(offset, groupID++);
+        }
+        // Materialize result.
+        wccLocalState->materialize(graph, groupArray, *table);
+    }
+
+    std::unique_ptr<GDSAlgorithm> copy() const override {
+        return std::make_unique<WeaklyConnectedComponent>(*this);
+    }
+
+private:
+    void findConnectedComponent(common::offset_t offset, int64_t groupID) {
+        visitedArray[offset] = true;
+        groupArray[offset] = groupID;
+        graph->scanNbrFwd(offset);
+        auto nbrs = graph->getNbrs();
+        for (auto nbrID : nbrs) {
+            if (visitedArray[nbrID.offset]) {
+                continue;
+            }
+            findConnectedComponent(nbrID.offset, groupID);
```

Comment:
I understand this is a poc for now but let's also add a todo to rewrite these algorithms to scale. This won't work at large scale.